experiments/without_metadata.py

```diff
@@ -0,0 +1,179 @@
+import logging
+from dataclasses import dataclass
+from typing import Optional
+
+from datasets import load_dataset
+from torch.utils.data import DataLoader
+
+import hydra
+from transformers import default_data_collator
+
+
+logger = logging.getLogger(__name__)
+
+
+def get_dataloaders(tokenizer, args):
```

Comment:
I have left this code for the moment as close as possible to the one used in the :hugs: causal language model training example scripts. 

In my opinion, we'll have two possibilities to include the addition of the metadata: 
- as pre-processing step in the manner of the `group_texts` method below;
- on-the-fly in a new `collate_fn` given to the `DataLoader`.

I would intuitively think that the on-the-fly solution is interesting because it would increase our pre-training dataset by putting some probability on the addition of metadata to each of the examples  (the same example could thus be increased differently from one time to another)

Reply:
The way [the function for adding metadata](https://github.com/bigscience-workshop/metadata/blob/8ef5bcb33fa3d4d785c862ba47649ca5dd75179a/metadata_utils.py#L25) works right now is similar to `group_texts` (i.e., with `map` and `batched=True`, see for example [how it is used here](https://github.com/bigscience-workshop/metadata/blob/master/test_metadata_utils.py#L80)).

After giving it some thought, I think a `collate_fn` might not be the right place for this: adding metadata can have a big impact on the number of training sequences we get from an input text. For example, if we have a text with 1,000 tokens and a maximum sequence length of 512, we can generate 2 training examples from it (with 512 and 488 tokens); but if the metadata we add as prefix to each example (e.g., URL, timestamp and topic) accounts for 200 tokens, it would be 4 training examples instead (with 312, 312, 312 and 64 tokens). I don't think this is something that should (can?) be handled by a `collate_fn`. But then again, I have very little experience with `DataLoader`, so I may be completely wrong here.

Reply:
Thanks a lot for the links and your explanation! :smile: 

Indeed, I agree with you that the easiest way is to make a function like `group_text`, I can even start plugging with your work if it's a good time, I think it's quite easy.

To bounce back on the `collate_fn`, you do raise an interesting point and I think there is a way to handle it in the `collate_fn` function. One way would be to draw with a certain probability equal to `1/(number of chunks associated with the text)` the part of the text to concatenate / augment with the metadata.

Finally, I think that choosing between a dataset mapping function (like `group_text`) or a `collate_fn` is a matter of choosing whether: 
- we prefer to save time on preprocessing operations or to save storage memory
- we want to have several variants for the same text example (with metadata or without metadata or with some metadata)
- we want to keep it simple